README.md

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Signed
by](https://img.shields.io/badge/Keybase-Verified-brightgreen.svg)](https://keybase.io/hrbrmstr)
![Signed commit
%](https://img.shields.io/badge/Signed_Commits-100%25-lightgrey.svg)
[![Linux build
Status](https://travis-ci.org/hrbrmstr/ggchicklet.svg?branch=master)](https://travis-ci.org/hrbrmstr/ggchicklet)
[![Coverage
Status](https://codecov.io/gh/hrbrmstr/ggchicklet/branch/master/graph/badge.svg)](https://codecov.io/gh/hrbrmstr/ggchicklet)
![Minimal R
Version](https://img.shields.io/badge/R%3E%3D-3.2.0-blue.svg)
![License](https://img.shields.io/badge/License-MIT-blue.svg)

# ggchicklet

Create Chicklet (Rounded Segmented Column) Charts

## Description

Sometimes it is useful to stylize column charts a bit more than just
bland rectangles. Methods are provided to create rounded rectangle
segmented column charts (i.e. “chicklets”).

## What’s Inside The Tin

  - `debates2019`: 2019-2020 U.S. Democratic Debate Candidate/Topic
    Times
  - `geom_chicklet`: Chicklet (rounded segmented column) charts

The following functions are implemented:

## Installation

``` r
install.packages("ggchicklet", repos = "https://cinc.rud.is")
# or
remotes::install_git("https://git.rud.is/hrbrmstr/ggchicklet.git")
# or
remotes::install_git("https://git.sr.ht/~hrbrmstr/ggchicklet")
# or
remotes::install_gitlab("hrbrmstr/ggchicklet")
# or
remotes::install_bitbucket("hrbrmstr/ggchicklet")
# or
remotes::install_github("hrbrmstr/ggchicklet")
```

NOTE: To use the ‘remotes’ install options you will need to have the
[{remotes} package](https://github.com/r-lib/remotes) installed.

## Usage

``` r
library(ggchicklet)

# current version
packageVersion("ggchicklet")
## [1] '0.5.2'
```

### From the NYTimes

``` r
library(hrbrthemes)
library(tidyverse)

data("debates2019")

debates2019 %>%
  filter(debate_group == 1) %>% 
  mutate(speaker = fct_reorder(speaker, elapsed, sum, .desc=FALSE)) %>%
  mutate(topic = fct_other(
    topic,
    c("Immigration", "Economy", "Climate Change", "Gun Control", "Healthcare", "Foreign Policy"))
  ) %>%
  ggplot(aes(speaker, elapsed, group = timestamp, fill = topic)) +
  geom_chicklet(width = 0.75) +
  scale_y_continuous(
    expand = c(0, 0.0625),
    position = "right",
    breaks = seq(0, 14, 2),
    labels = c(0, sprintf("%d min.", seq(2, 14, 2)))
  ) +
  scale_fill_manual(
    name = NULL,
    values = c(
      "Immigration" = "#ae4544",
      "Economy" = "#d8cb98",
      "Climate Change" = "#a4ad6f",
      "Gun Control" = "#cc7c3a",
      "Healthcare" = "#436f82",
      "Foreign Policy" = "#7c5981",
      "Other" = "#cccccc"
    ),
    breaks = setdiff(unique(debates2019$topic), "Other")
  ) +
  guides(
    fill = guide_legend(nrow = 1)
  ) +
  coord_flip() +
  labs(
    x = NULL, y = NULL, fill = NULL,
    title = "How Long Each Candidate Spoke",
    subtitle = "Nights 1 & 2 of the June 2019 Democratic Debates",
    caption = "Each bar segment represents the length of a candidate’s response to a question.\n\nOriginals <https://www.nytimes.com/interactive/2019/admin/100000006581096.embedded.html?>\n<https://www.nytimes.com/interactive/2019/admin/100000006584572.embedded.html?>\nby @nytimes Weiyi Cai, Jason Kao, Jasmine C. Lee, Alicia Parlapiano and Jugal K. Patel\n\n#rstats reproduction by @hrbrmstr"
  ) +
  theme_ipsum_rc(grid="X") +
  theme(axis.text.x = element_text(color = "gray60", size = 10)) +
  theme(legend.position = "top")
```

<img src="man/figures/README-nyt-1.png" width="100%" />

## ggchicklet Metrics

| Lang | \# Files |  (%) |   LoC |  (%) | Blank lines |  (%) | \# Lines |  (%) |
| :--- | -------: | ---: | ----: | ---: | ----------: | ---: | -------: | ---: |
| HTML |       12 | 0.55 | 10200 | 0.92 |        3169 | 0.97 |      501 | 0.71 |
| R    |        8 | 0.36 |   788 | 0.07 |          45 | 0.01 |      122 | 0.17 |
| Rmd  |        2 | 0.09 |   117 | 0.01 |          45 | 0.01 |       85 | 0.12 |

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.


NEWS.md
0.5.2
* Data updated for latest debate (2020-02-25)

0.5.1
* Data updated for latest debate (2020-02-19)

0.5.0
* Data updated for latest debate (2020-02-07)

0.4.0
* Data updated for latest debate

0.1.0 
* Initial release


tests/test-all.R
library(testthat)

test_check("ggchicklet")


tests/tinytest.R

if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("ggchicklet")
}



inst/tinytest/test_ggchicklet.R
library(ggplot2)
library(ggchicklet)

data("debates2019")

spkr_ordr <- aggregate(elapsed ~ speaker, data = debates2019, sum)
spkr_ordr <- spkr_ordr[order(spkr_ordr[["elapsed"]]),]

debates2019$speaker <- factor(debates2019$speaker, spkr_ordr$speaker)

ggplot(debates2019) +
  geom_chicklet(aes(speaker, elapsed, group = timestamp, fill = topic)) -> gg

print(gg)

ggplot_build(gg) -> gb

ggplot_gtable(gb) -> gt

expect_true(
  all(c("GeomChicklet", "GeomRrect") %in% class(gb$plot$layers[[1]]$geom))
)



R/datasets.R
#' @md
#' @title 2019-2020 U.S. Democratic Debate Candidate/Topic Times
#' @description The New York Times and other media outlets kept track of the time each
#'              candidate spent talking including the timestamp of the start of the blathering
#'              and the topic up for debate. This dataset only includes candidates and
#'              topic times. The complete datasets (See References) also include moderator
#'              metadata and opening/closing statement records.
#' @format data frame with columns: `elapsed` (dbl), `timestamp` (time), `speaker` (chr), `topic` (chr)
#'         `debate_date` (date), `debate_group` (dbl), `night` (dbl)
#' @docType data
#' @keywords datasets
#' @name debates2019
#' @references <https://www.nytimes.com/interactive/2019/06/26/us/elections/debate-speaking-time.html>
#' @references <https://www.nytimes.com/interactive/2019/06/27/us/elections/debate-speaking-time.html>
#' @references <https://www.nytimes.com/interactive/2019/07/30/us/elections/debate-speaking-time.html>
#' @references <https://www.nytimes.com/interactive/2019/07/31/us/elections/debate-speaking-time.html>
#' @references <https://www.nytimes.com/interactive/2019/09/12/us/elections/debate-speaking-time.html>
#' @references <https://www.nytimes.com/interactive/2019/10/15/us/elections/debate-speaking-time.html>
#' @usage data("debates2019")
NULL

R/a-geom-rect.R
geom_rrect <- function(mapping = NULL, data = NULL,
                       stat = "identity", position = "identity",
                       radius = grid::unit(6, "pt"),
                       ...,
                       na.rm = FALSE,
                       show.legend = NA,
                       inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomRrect,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      radius = radius,
      na.rm = na.rm,
      ...
    )
  )
}

GeomRrect <- ggplot2::ggproto(
  "GeomRrect", ggplot2::Geom,

  default_aes = ggplot2::aes(
    colour = NA, fill = "grey35", size = 0.5, linetype = 1, alpha = NA
  ),

  required_aes = c("xmin", "xmax", "ymin", "ymax"),

  draw_panel = function(self, data, panel_params, coord,
                        radius = grid::unit(6, "pt")) {

    coords <- coord$transform(data, panel_params)

    lapply(1:length(coords$xmin), function(i) {

      grid::roundrectGrob(
        coords$xmin[i], coords$ymax[i],
        width = (coords$xmax[i] - coords$xmin[i]),
        height = (coords$ymax[i] - coords$ymin)[i],
        r = radius,
        default.units = "native",
        just = c("left", "top"),
        gp = grid::gpar(
          col = coords$colour[i],
          fill = alpha(coords$fill[i], coords$alpha[i]),
          lwd = coords$size[i] * .pt,
          lty = coords$linetype[i],
          lineend = "butt"
        )
      )

    }) -> gl

    grobs <- do.call(grid::gList, gl)

    ggname("geom_rrect", grid::grobTree(children = grobs))

  },

  draw_key = ggplot2::draw_key_polygon

)

R/geom-chicklet.R
#' Chicklet (rounded segmented column) charts
#'
#' This geom behaves much like [ggplot2::geom_col()] but provides the option to
#' set a corner radius to turn sharp-edged bars into rounded rectangles; it also
#' sets some sane defaults for making chicklet charts.
#'
#' \if{html}{
#' A sample of the output from \code{geom_chicklet()}:
#'
#' \figure{README-nyt-1.png}{options: width="100\%" alt="Figure: README-nyt-1.png"}
#' }
#'
#' \if{latex}{
#' A sample of the output from \code{geom_chicklet()}:
#'
#' \figure{README-nyt-1.png}{options: width=10cm}
#' }
#'
#' @section Aesthetics:
#' `geom_chicklet()` understands the following aesthetics:
#' - `x`
#' - `y`
#' - `alpha`
#' - `colour`
#' - `fill`
#' - `group`
#' - `linetype`
#' - `size`
#'
#' Use both `fill` and `group` when you want a fill colour per-segment but
#' want to order the segments by another column (as in the Examples).
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_point
#' @inheritParams ggplot2::geom_col
#' @param radius corner radius (default 3pt)
#' @note the chicklet/segment stack positions are default set to be reversed (i.e.
#'       left-to-right/bottom-to-top == earliest to latest order).
#' @export
#' @rdname geom_chicklet
#' @examples
#' library(ggplot2)
#'
#' data("debates2019")
#'
#' # set the speaker order
#' spkr_ordr <- aggregate(elapsed ~ speaker, data = debates2019, sum)
#' spkr_ordr <- spkr_ordr[order(spkr_ordr[["elapsed"]]),]
#'
#' debates2019$speaker <- factor(debates2019$speaker, spkr_ordr$speaker)
#'
#' ggplot(debates2019) +
#'   # use 'group' to control left-to-right order
#'   geom_chicklet(aes(speaker, elapsed, group = timestamp, fill = topic)) +
#'   scale_y_continuous(expand = c(0, 0.01), position = "right") +
#'   coord_flip() +
#'   labs(x = NULL, y = "Minutes Spoken", fill = NULL) +
#'   theme_minimal() +
#'   theme(panel.grid.major.y = element_blank()) +
#'   theme(legend.position = "bottom")
geom_chicklet <- function(mapping = NULL, data = NULL,
                          position = ggplot2::position_stack(reverse = TRUE),
                          radius = grid::unit(3, "pt"), ..., width = NULL,
                          na.rm = FALSE, show.legend = NA, inherit.aes = TRUE) {
  layer(
    data = data, mapping = mapping, stat = "identity",
    geom = GeomChicklet, position = position, show.legend = show.legend,
    inherit.aes = inherit.aes, params = list(
      width = width, radius = radius, na.rm = na.rm, ...
    )
  )
}

draw_key_rrect <- function(data, params, size) { # nocov start
  grid::roundrectGrob(
    r = min(params$radius, unit(3, "pt")),
    default.units = "native",
    width = 1, height = 0.6,
    name = "lkey",
    gp = grid::gpar(
      col = params$color %l0% "white",
      fill = alpha(data$fill %||% data$colour %||% "grey20", data$alpha),
      lty = data$linetype %||% 1
    )
  )
} # nocov end

#' @rdname geom_chicklet
#' @export
GeomChicklet <- ggproto( # nocov start
  "GeomChicklet", GeomRrect,

  required_aes = c("x", "y"),

  default_aes = ggplot2::aes(
    colour = "white", fill = "grey35", size = 0.5, linetype = 1, alpha = NA
  ),

  non_missing_aes = c("xmin", "xmax", "ymin", "ymax"),

  setup_data = function(data, params) {
    data$width <- data$width %||%
      params$width %||% (resolution(data$x, FALSE) * 0.9)
    transform(data,
              ymin = pmin(y, 0), ymax = pmax(y, 0),
              xmin = x - width / 2, xmax = x + width / 2, width = NULL
    )
  },

  draw_panel = function(self, data, panel_params, coord,width = NULL, radius = grid::unit(3, "pt")) {
    ggproto_parent(GeomRrect, self)$draw_panel(data, panel_params, coord, radius = radius)
  },

  draw_key = draw_key_rrect

) # nocov end
